staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -331,6 +409,16 @@ func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiService *v
 func (s *specAggregator) RemoveAPIServiceSpec(apiServiceName string) error {
 	s.rwMutex.Lock()
 	defer s.rwMutex.Unlock()
+	if utilfeature.DefaultFeatureGate.Enabled(features.OpenAPIV3) {
+		s.openAPIV3rwMutex.Lock()
```

Comment:
Should we lock both mutex here? Any chance we could get into a deadlock situation?

Reply:
The v2 and v3 locks are for the v2 and v3 spec map respectively and are independent of each other. There isn't another place in the code where both locks are used so deadlock shouldn't be possible at the moment. We could do the writes one after the other to prevent both locks from being held at the same time but we won't be able to use the `defer` anymore.

Reply:
The section at the beginning is only taking care of v3 and the section at the end is taking care of v2, can't you split in two? Also, I'd probably still use defer by splitting into two functions? I think this shows how the Aggregator might be better as two separate aggregator here.

Reply:
Completed refactored per sttts's suggestion.